### PR TITLE
[Asset] set version automatically from git commit hash

### PR DESCRIPTION
**Pull request by ewgRa**
| Q | A |
| --- | --- |
| Branch? | "master" |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR |  |

There is old idea to implement auto versions for assets without external tools like "sed" and so on on build stage. This PR is a _concept_ for get version from latest git commit hash.

I think this strategy will cover most of Symfony deployment cycles. Also there is can be another strategies, like get directory timestamp, tag name and so on.

Usage is simple:

1) Register compiler pass: 

``` php
class AppBundle extends Bundle
{
    public function build(ContainerBuilder $container)
    {
        parent::build($container);
        $container->addCompilerPass(new GitCommitHashCompilerPass());
    }
}
```

2) Create version strategy: 

``` yaml
    assets.version.strategy.gitcommithash:
        class: Symfony\Component\Asset\VersionStrategy\StaticVersionStrategy
        arguments: ["%git_commit_hash_version_strategy%"]
```

3) Set strategy:

``` yaml
    assets:
        version_strategy: assets.version.strategy.gitcommithash
```

If idea will be accepted in general, I will add tests and try to make it more general to reduce number of steps that needed to enable it.

### CODE REVIEW

File: src/Symfony/Component/Asset/CompilerPass/GitCommitHashCompilerPass.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+     */
+    public function process(ContainerBuilder $container)
+    {
+        $process = new Process('git log -1');
+
+        if (0 !== $process->run()) {
+            throw new \Exception('Git command return wrong exit code');
+        }
+
+        $container->setParameter('git_commit_hash_version_strategy', sha1($process->getOutput()));
```
Comment (stof) [inline src/Symfony/Component/Asset/CompilerPass/GitCommitHashCompilerPass.php]: getting a sha1 of the logs is weird. Why not getting the commit hash instead ?

File: src/Symfony/Component/Asset/CompilerPass/GitCommitHashCompilerPass.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Asset\CompilerPass;
```
Comment (stof) [inline src/Symfony/Component/Asset/CompilerPass/GitCommitHashCompilerPass.php]: this should not be in the Asset component, as there is nothing related to the Asset component in this code (there is code related to the DependencyInjection and Process components)

File: src/Symfony/Component/Asset/CompilerPass/GitCommitHashCompilerPass.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    /**
+     * @param ContainerBuilder $container
+     *
+     * @throws \Exception
+     */
+    public function process(ContainerBuilder $container)
+    {
+        $process = new Process('git log -1');
+
+        if (0 !== $process->run()) {
```
Comment (stof) [inline src/Symfony/Component/Asset/CompilerPass/GitCommitHashCompilerPass.php]: we have ``$process->mustRun()`` which triggers a much better exception (giving more details about what is wrong)

File: src/Symfony/Component/Asset/CompilerPass/GitCommitHashCompilerPass.php (hunk 1 above)
Comment (stof) [inline src/Symfony/Component/Asset/CompilerPass/GitCommitHashCompilerPass.php]: the parameter name ``git_commit_hash_version_strategy`` is weird btw. This is not a strategy at all.

**Comment by stof** [comment]
I'm not sure this should be in core. Bumping the asset version for each commit is the poor man solution, which will do cache busting far too often.

**Comment by javiereguiluz** [comment]
Maybe it would be better to use `hash_init()`, `hash_update()`, etc. to build incrementally the hash of the entire web asset files. (I learned this trick from @lyrixx).

**Comment by ewgRa** [comment]
@javiereguiluz Can you give link or details? It is interesting idea, but I fear about performance and IO load. I think it can be another one PassCompiler. For now question is about can such strategies be in core, or not.

@stof Since this will be not by default, developers can decide, is this acceptable or not and choose best auto version strategy for project. We have not so small project, but usually we have 1-2 release every week and this refresh is not so critical for us. And I think a lot of projects have same case.

**Comment by javiereguiluz** [comment]
The PHP [hash_init](http://php.net/hash_init), [hash_update_file](http://php.net/hash_update_file) and [hash_final](http://php.net/hash_final) allow you to create for example the global `md5` hash of all the different files without having to combine the files or concatenate different hashes or any other dirty hack. 

And about performance, you are right. This cannot be used to generate the version dynamically. This only makes sense when generating the version on each deploy, when assets won't change anymore.

**Comment by lyrixx** [comment]
In order to fix this issue, we have a jenkins jobs that compute the asset version and we store this information in a git note. So for each commit we have the matching asset version.

**Comment by stof** [comment]
@javiereguiluz building such a strategy should be discussed elsewhere than in this PR

File: src/Symfony/Component/Asset/CompilerPass/GitCommitHashCompilerPass.php (hunk 1 above)
Comment (ewgRa) [inline src/Symfony/Component/Asset/CompilerPass/GitCommitHashCompilerPass.php]: > getting a sha1 of the logs is weird. Why not getting the commit hash instead ?

Because for this parse output needed and result will be same - some unique hash of commit. Less work and no dependency on git output format, I fear that it can be reconfigured.

Why it looks weird? You think have real git commit hash looks better and have advantages?

**Comment by ewgRa** [comment]
Review fixes done.

Actually I think, what if in CompilerPass set container definition for assets.version.strategy.gitcommithash service? Less steps to enable asset strategy.

File: src/Symfony/Component/Asset/CompilerPass/GitCommitHashCompilerPass.php (hunk 2 above)
Comment (ewgRa) [inline src/Symfony/Component/Asset/CompilerPass/GitCommitHashCompilerPass.php]: I just think that because components can be used independently, there is Asset component looks better for this. Also as if this PR will be accepted, there will be more strategies, and all this stuff is not related to DI component.

**Comment by ewgRa** [comment]
Also would be nice to have only one step to enable such strategies, just set version_strategy: assets.version.strategy.gitcommithash, and there is no any additional changes needed, like define service, add CompilerPass.

It will be very friendly for developers and cover a lot of deployment cases without third-party solutions.

**Comment by nicolas-grekas** [comment]
What about adding a new `GitVersionStrategy` instead? For generating the version, I suggest using `git rev-parse HEAD`

**Comment by ewgRa** [comment]
@nicolas-grekas agree with git rev-parse, looks much better. Commit done.

What is a point to create new GitVersionStrategy? I try to check which one advantages it gives, but don't see any. As far as I understand, service will be still looks like:

```
assets.version.strategy.gitcommithash:
        class: Symfony\Component\Asset\VersionStrategy\GitVersionStrategy
        arguments: ["%git_commit_hash%"]
```

and user will need to define it himself and know this "git_commit_hash" parameter. Right?

I try to check how it can be declared in FrameworkBundle for exmaple, but it can be only as non public service, or for example as abstract with hardcoded checks in FrameworkExtension. Similar to StaticVersionStrategy definition, but with Git strategy there is also one parameter, that somehow must be ignored if developer do not use this strategy.

Ideal use case is just define
```
    assets:
        version_strategy: assets.version.strategy.gitcommithash
```
and another steps will be done automatically (service definition, compiler pass). But don't see how to achieve it with current code, is there any examples with exists code?

**Comment by nicolas-grekas** [comment]
@ewgRa: component responsibility. E.g. if you add a GitVersionStrategy, users will find it easily. But as a comiler pass (for di extension), it's not self explanatory.
(btw in your example above, there would be no need for any argument/param since it would be the strategy's responsibility to get it).

**Comment by nicolas-grekas** [comment]
oh, and I'd suggest then adding a private service in FrameworkBundle, called `assets.git_version_strategy`
so that the only configuration you'd had to do would be:
`
    assets:
        version_strategy: assets.git_version_strategy
`

**Comment by ewgRa** [comment]
@nicolas-grekas Done. It is how I see it.

For now it is enough just set version_strategy: assets.git_version_strategy and it will be work without any additional changes.

Did you see what can be improved here? I don't like "removeDifinition", but don't see better way check that developer really use this functionality and we don't need to process service argument.

$format in GitVersionStrategy I left for developers who want another format. They just will be needed to redefine in their services:

```
    assets.git_version_strategy:
        class: Symfony\Component\Asset\VersionStrategy\GitVersionStrategy
        arguments: ["", '%%s?custom_%%s']
```

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 4]
```diff
@@ -574,10 +574,16 @@ private function registerAssetsConfiguration(array $config, ContainerBuilder $co
     {
         $loader->load('assets.xml');
 
+        $removeGitStrategy = true;
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: this logic is not needed: since the service is private, it will be removed automatically when the container is compiled

File: src/Symfony/Component/DependencyInjection/Compiler/GitCommitHashPass.php [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    /**
+     * @param ContainerBuilder $container
+     */
+    public function process(ContainerBuilder $container)
+    {
+        if (!$container->hasDefinition('assets.git_version_strategy')) {
+            return;
+        }
+
+        $process = new Process('git rev-parse HEAD', $container->getParameter('kernel.root_dir'));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/GitCommitHashPass.php]: I think this computation should be done inside the GitVersionsStategy. This "pass" is not needed.

Comment (ewgRa) [inline src/Symfony/Component/DependencyInjection/Compiler/GitCommitHashPass.php]: Than it will be calculated for every request? Right? This is what I want to avoid. Do not exec this for every request.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 4 above)
Comment (ewgRa) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: This logic is for developers who not use this strategy at all.

They can have git not installed, or not properly php configured to exec Process functionality and so on. They will be very disappointed if GitHash compiler will try to get git hash for them.

And this logic just say to GitHashCompilerPass - "hey, developer do not use this functionality, just skip it". 

If I will remove this logic - how GitHashCompilerPass can understand that this hash is not needed to be processed?

File: src/Symfony/Component/DependencyInjection/Compiler/GitCommitHashPass.php (hunk 5 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/GitCommitHashPass.php]: Good point, still, the current logic is assuming to much (the service name convention esp...)

Comment (ewgRa) [inline src/Symfony/Component/DependencyInjection/Compiler/GitCommitHashPass.php]: Another CompilerPass'es from FrameworkBundle do the same, check hasDefinition and return if there is no such definition. Also as there is one removeDefinition already there.

Comment (ewgRa) [inline src/Symfony/Component/DependencyInjection/Compiler/GitCommitHashPass.php]: So, did you see better way?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 4 above)
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: So, let's reverse this logic and define dynamically the `assets.git_version_strategy` service when it's the value of the `version_strategy` config option. And also, let's compute the git hash inline in the FrameworkExtension. (which means removing GitVersionStrategy and the patch on the xml file).
Btw, I think the git process should have its working directory forced to `kernel.root_dir`.

Comment (ewgRa) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: > Btw, I think the git process should have its working directory forced to kernel.root_dir.

It is already implemented:
```$process = new Process('git rev-parse HEAD', $container->getParameter('kernel.root_dir'));```

Ok, I will do another changes. But actually don't see why it is better than current implementation, just will be in one place, but not so flexible on my mind.

**Comment by lyrixx** [comment]
Hi

> I'm not sure this should be in core. Bumping the asset version for each commit is the poor man solution, which will do cache busting far too often.

I totally agree with @stof ; 
This solution is really too cheap and it's not a best practice. So it should not belong to Symfony

**Comment by ewgRa** [comment]
@lyrixx what is a best practice in this case?

Manually update asset_version? It is pain to always think about this.

**Comment by nicolas-grekas** [comment]
@ewgRa compute the version based on actual changes of any assets (which can be automated by e.g. a deployment script)

**Comment by lyrixx** [comment]
We have a in house solution for that. It's not perfect but it's better. I already commented [here](https://github.com/symfony/symfony/pull/18642#issuecomment-214707333) but I gonna explain a bit more.

We have a jenkins jobs that is executed for each merge the `master` branch  (preprod) and the `prod` branch (prod).

This job run `composer install && gulp && app/console upload-assets`.

"upload assets" performs fews things:

* It compute a hash based on the content of each file that live our `assets` folders (.css, .less. .js .jsx ...). We use `hash_init()`, `hash_update()`, ... PHP function for that
* It store the this hash in a git commit note
* If assets are not yet on AWS S3 (our CDN for assets (with CloudFront)) it uploads all assets to S3

Then during a deploy we get the related asset version form the current git commit note. And we add this value to the `env` for the `composer install` task. So finally the value is in the Symfony DIC and here we go.

Obviously everything is automated and we never have to worry about changing assets version, uploading assets, cache bursting etc.

Finally this version is not perfect because if we change only one css for example, we invalidate all the cache...

But it's really really more efficient than this PR which invalidate cache for each commit and so each deploy (we deploy many times per day like many others...)

**Comment by stof** [comment]
> Finally this version is not perfect because if we change only one css for example, we invalidate all the cache...

A better solution would be to hash each file separately, and then generating a json (or whatever other format which can store a hashmap and be decoded easily in PHP) file with the mapping between assets and their version. Then, a strategy implementation could perform a lookup in this map to find the version for each asset.
This would allow to invalidate only the asset which changed.
This is much more efficient when some assets are changing much more often than others.

**Comment by ewgRa** [comment]
You can close it as you think it is not a core functionality, perhaps I will just implement this in Bundle.

Since I haven't poll results like "95% of Symfony installations update its asset version manually and use deploy process like 'git checkout newtag'" I haven't idea how to argue that it would be nice to have it in core.

@stof I think about have hash from assetic compiler or similar. But not all developers use assetic, not all developers have one and standard source of assets and it will be "another one strategy" that hard to say how used it will be.

**Comment by stof** [comment]
@ewgRa writing such a strategy is something I have in mind. I haven't worked on it yet, but I start to have an idea about how it could work (without the need to use Assetic btw, it could work for any assets as long a they are available in the codebase). Once I write it, I will make it available open-source

**Comment by ewgRa** [comment]
@stof Idea is "inject" some code on "asset" twig template compilation?


Ok, than I will close it for now, since there will be better solution sooner or later.
